IRSA-816:Add processor to generate JSON for IRSA (IBE) from imageset master-table

lznakano opened the pull request:
Add ImageSetConverter

To test it:

1. built firefly
2. http://localhost:8080/firefly/firefly-dev.html
3. select image-concept

ejoliet commented:
Nice! I see SEIP in the panel now! Great!
Although the conversion seems to be working nicely (i will look at the code later), i see couple of issues that needs attention, either by fixing it in this ticket or raising separate ticket(s) if you feel it has to be tackle outside by someone else.
- checkbox not working (search not working as a consequence): there are errors thrown in the console 

```Warning: flattenChildren(...): Encountered two children with the same key, `1:$undefined`. Child keys must be unique; when two children share a key, only the first child will be used.
    in div (created by CheckboxGroupInputFieldView)
    in CheckboxGroupInputFieldView (created by FGConnector(CheckboxGroupInputFieldView))
    in FGConnector(CheckboxGroupInputFieldView) (created by FilterSelect)
    in div (created by FilterSelect)
    in FilterSelect (created by FilterPanelView)
    in div (created by CollapsiblePanel)```
I suspect some kind of mapping label to value not consistent ( @loitly might help you).

If i merge @loitly's branch in yours i can see this working out of the box! FANTASTIC job both!

I see an issue though, but we can wait when both of you merges:
- Panel jump outside the form's frame: probably need to fix the width or play with overflow to have an horizontal scroll bar. Some datasets (i.e. SEIP) have more bands and longer labels affect it also, so the line of checkboxes take lot of space horizontally. To be discussed later.

ejoliet commented on src/firefly/java/edu/caltech/ipac/firefly/resources/master-table.csv:
```diff
@@ -0,0 +1,51 @@
+name,missionLabel (project),missionId,dataProductLabel,dataProductId,acronym,instrumentId,instrumentLabel,wavebandId,wavebandDesc,wavelength (in microns),wavelengthDesc,projectTypeKEy,projectTypeDesc,minSizeDeg,maxSizeDeg,helpUrl,tooltip,apiType,atlasTable,dataType
```
Do we need 2 master-table for IRSA?

ejoliet commented on src/firefly/java/edu/caltech/ipac/firefly/server/visualize/imagesources/IrsaMasterDataSource.java:
```diff
@@ -27,18 +28,20 @@
+            //String dataStr= URLDownload.getStringFromURL(url,null);
+           //JSONParser parser= new JSONParser();
+            //JSONArray data= (JSONArray)parser.parse(dataStr);
+            ImageSetConverter ic = new ImageSetConverter("IRSA");
```
Please, make "IRSA" a constant and call it from the constructor.

lznakano replied:
OK

ejoliet commented on src/firefly/java/edu/caltech/ipac/firefly/resources/master-table.csv:
```diff
@@ -0,0 +1,51 @@
+name,missionLabel (project),missionId,dataProductLabel,dataProductId,acronym,instrumentId,instrumentLabel,wavebandId,wavebandDesc,wavelength (in microns),wavelengthDesc,projectTypeKEy,projectTypeDesc,minSizeDeg,maxSizeDeg,helpUrl,tooltip,apiType,atlasTable,dataType
+JSON DM-12001,project,,subProject,surveyKey,,,,surveyKeyBand,title,,,,,minRangeDeg,maxRangeDeg,,,Service,surveyKeyAlt,
```
those 2 lines can be removed. Also the code associate to it. See comment on the class.

lznakano replied:
You looked at the wrong table.  The one you looked at is the one you made.  The table used is irsa-master-table.csv.  I can delete the old one before I merge to the dev.

ejoliet replied:
yes, please, remove it before merge.

ejoliet commented on src/firefly/java/edu/caltech/ipac/firefly/util/ImageSetConverter.java:
```diff
@@ -0,0 +1,209 @@
+        imageSetToJSONArray = createImageJSONArray(masterTable);
+        //test the result
+     /*   try {
+            writeJSONArrayToFile(imageSetToJSONArray,  "/Users/zhang/IRSA_Dev/testingData/irsa-master-table.json");
```
this can be removed before merging.

lznakano replied:
Yes,  good idea.  I will move it to a test case.

loitly commented on src/firefly/java/edu/caltech/ipac/firefly/server/util/DsvToDataGroup.java:
```diff
@@ -33,14 +27,53 @@
+        return getData(reader, format);
+
+       /* List<DataType> columns = new ArrayList<DataType>();
+        CSVParser parser = new CSVParser(reader, format);
```
You should go ahead and remove this block of commented code.

lznakano replied:
OK

loitly commented on src/firefly/java/edu/caltech/ipac/firefly/server/visualize/imagesources/IrsaMasterDataSource.java:
```diff
@@ -27,18 +28,20 @@
+            //URL url= ImageMasterData.class.getResource("/edu/caltech/ipac/firefly/resources/irsa-image-data.json");
+            //String dataStr= URLDownload.getStringFromURL(url,null);
+           //JSONParser parser= new JSONParser();
+            //JSONArray data= (JSONArray)parser.parse(dataStr);
```
Should remove these commented lines as well.

loitly commented on src/firefly/java/edu/caltech/ipac/firefly/util/ImageSetConverter.java:
```diff
@@ -0,0 +1,209 @@
+ * This class will read in the ImageSet information table (maybe from an API) and
+ * the convert to a JSONObject.
+ */
+public class ImageSetConverter {
```
A couple of things about this class.
- should not be responsible for converting into JSONObject.  should only create  List<ImageMasterDataEntry>
- why is it in a util package?  nothing generic about this class.  it's an extension of IrsaMasterDataSource.  should be inside IrsaMasterDataSource.java or in the same package.  if it's in a separate file, it should have 'irsa' in its name as well.

ejoliet replied:
So, you are saying that this should be the 'Combiner' and another class should convert the combined entries into json, is that what i hear? I agree.

lznakano replied:
Emmanuel suggested to make it as a utility class.  Where should I put now?  Please give suggestions!  Thanks!

robyww replied:
I don't understand this class either. JSON is the transport layer that is handled laster by ImageMasterData.  This point of this class should it create a `List<ImageMasterDataEntry>`.  There should be no json involved.

ejoliet replied:
The ticket was misleading because it had 2 ideas in one but you did part of it by using json directly in the server-side. Between Loi and I, i think we've explained the differences to Lijun and the (new) approach to keep the java side to handle the setters of the values from the master table (IRSA specific btw). The idea behind the ticket was to have a parser to generate objects that would end into json in the client. The only part needed is the parser into java objects that can be sent to client on server (command) request. 
Does it make more sense now? @lznakano

loitly commented on src/firefly/js/visualize/ui/MultiItemViewerView.jsx:
```diff
@@ -283,7 +283,7 @@ function computeTitleYOffset(sparseGridTitleLocation, itemPosTitle,rowIdx) {
 }
 function computeDataYOffset(sparseGridTitleLocation, itemPosTitle,rowIdx) {
-    let titleOffset= sparseGridTitleLocation==='left' && itemPosTitle[0].hasData ? 0 : 30;
+    let titleOffset= sparseGridTitleLocation==='left' && itemPosTitle[0] && itemPosTitle[0].hasData ? 0 : 30;
```
`itemPosTitle[0] && itemPosTitle[0].hasData`
can be replaced with 
get(itemPosTitle, '0.hasData')

ejoliet commented on src/firefly/java/edu/caltech/ipac/firefly/util/ImageSetConverter.java:
```diff
@@ -0,0 +1,209 @@
+                case "string":
+                    String val = obj==null? null:((String) obj).trim();
+
+                    if (dataTypes[i].getKeyName().equalsIgnoreCase("missionLabel (project)")){
```
I think we need to talk about this. The mapping and having the hard-coded values is fine for now but it really should come from metadata API (future) or at least from properties so we don't have to change this code to make it work. Let's discuss this together.

ejoliet commented on src/firefly/java/edu/caltech/ipac/firefly/util/ImageSetConverter.java:
```diff
@@ -0,0 +1,209 @@
+        for(int i=0; i<params.size(); i++){
+            plotReqParams.put(keys[i], params.get(keys[i]));
+        }
+        jsonObject.put( "plotRequestParams",plotReqParams );
```
Same comment as above.

lznakano replied:
@ejoliet This is what you suggested at the first place: We hard coded the mapping now and later we will get them from the API.

ejoliet replied:
yes, is fine for now. But what i meant for API, is not related to the mapping itself, it needs to be 'hardcoded' outside because if API change (metadata columns), your code should be the same, the mapping will change. It could be read from props, but as i said, is fine for now. When we get to the point of getting it from API, we can move the mapping out.

ejoliet commented on src/firefly/java/edu/caltech/ipac/util/IpacTableUtil.java:
```diff
@@ -182,6 +182,7 @@ public static void guessFormatInfo(DataType dataType, String value) {
 
     public static void guessFormatInfo(DataType dataType, String value, int precision) {
 
+        if (StringUtils.isEmpty(value)) return; //LZ added IRSA-816
```
GOOD CATCH!

ejoliet commented on src/firefly/js/visualize/ui/MultiItemViewerView.jsx:
```diff
@@ -283,7 +283,7 @@ function computeTitleYOffset(sparseGridTitleLocation, itemPosTitle,rowIdx) {
 }
 function computeDataYOffset(sparseGridTitleLocation, itemPosTitle,rowIdx) {
-    let titleOffset= sparseGridTitleLocation==='left' && itemPosTitle[0].hasData ? 0 : 30;
+    let titleOffset= sparseGridTitleLocation==='left' && itemPosTitle[0] && itemPosTitle[0].hasData ? 0 : 30;
```
Why 'hasData' can't also check if undefined or not? Then no need to have both in the boolean evaluation, right?

lznakano replied:
@loi 
Can you explain the advantage of doing that way?  I know that I can do that and Emmanuel and also talked about it.  But I think using Map is  the same and have less hard coded key names.

loitly replied:
`get` is a lodash function.  It will handle `null` cases as well as defaults.  It's especially useful when you're fetching a deep object.  I am suggesting it in case you're not aware.  If you are and prefer your way, it's fine.  That's why it was a suggestion.

in your example, 
```
itemPosTitle[0] && itemPosTitle[0].hasData
```
it's only 2 level deep.  imagine 4 levels.. then it would look like this...
```
itemPosTitle[0] && itemPosTitle[0].main && itemPosTitle[0].main[0] && itemPosTitle[0].main[0].hasData
```
in this case, you can replace it with:
```
get(itemPosTitle, '0.main.0.hasData')
```
or, 
```
get(itemPosTitle, '0.main.0.hasData', false)
```
if you wanted the returned value to be a boolean type.

loitly commented:
Please do the following:
- remove any references of lsst in ImageSetConverter.java
- remove file ImageSetConverter.java and move logic into IrsaMasterDataSource.java
-in createDataList(),  code should look like this:

```
    public List createDataList(String masterTable) throws IOException, FailedRequestException {

        DataGroup inDg = getDataFromMasterTable(masterTable);
        List<ImageMasterDataEntry>  retList= new ArrayList<>();

        for(DataObject row : inDg.values()) {
            ImageMasterDataEntry entry = new ImageMasterDataEntry();
            entry.setProject(String.valueOf(row.getDataElement("dataProductId")));
            entry.setSubProject(String.valueOf(row.getDataElement("instrumentId"));
                ...

            retList.add( entry);
        }
        return retList;
    }
```

lznakano commented:
Sorry, I put the comment in a wrong place.  I mean this block.


Thanks!


On 10/2/17 3:42 PM, Loi Ly wrote:
>
> Please do the following:
>
>   * remove any references of lsst in ImageSetConverter.java
>   * remove file ImageSetConverter.java and move logic into
>     IrsaMasterDataSource.java
>     -in createDataList(), code should look like this:
>
> |public List createDataList(String masterTable) throws IOException, 
> FailedRequestException { DataGroup inDg = 
> getDataFromMasterTable(masterTable); List<ImageMasterDataEntry> 
> retList= new ArrayList<>(); for(DataObject row : inDg.values()) { 
> ImageMasterDataEntry entry = new ImageMasterDataEntry(); 
> entry.setProject(String.valueOf(row.getDataElement("dataProductId"))); 
> entry.setSubProject(String.valueOf(row.getDataElement("instrumentId")); 
> ... retList.add( entry); } return retList; } |
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Caltech-IPAC/firefly/pull/469#issuecomment-333686016>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ALbTfRhlI_fjo7UpJ7fGywDmnTwL21Sxks5soWblgaJpZM4PrAhp>.
>

-- 
Lijun Zhang

LSST/IPAC
California Institute of Technology
770 S. Wilson Ave, Mail Code 100-22
Pasadena, CA 91125, USA
Voice: 1-626-395-1972 Fax:1-626-395-1816

robyww commented on src/firefly/java/edu/caltech/ipac/firefly/resources/master-table.csv:
```diff
@@ -0,0 +1,51 @@
+name,missionLabel (project),missionId,dataProductLabel,dataProductId,acronym,instrumentId,instrumentLabel,wavebandId,wavebandDesc,wavelength (in microns),wavelengthDesc,projectTypeKEy,projectTypeDesc,minSizeDeg,maxSizeDeg,helpUrl,tooltip,apiType,atlasTable,dataType
+JSON DM-12001,project,,subProject,surveyKey,,,,surveyKeyBand,title,,,,,minRangeDeg,maxRangeDeg,,,Service,surveyKeyAlt,
+string,string,string,string,string,string,string,,string,string,float,string,,,float,float,,string,string,string,string
```
i don't think line 2 and 3 should be in the CSV.  There should only be a title line.

robyww commented on src/firefly/java/edu/caltech/ipac/firefly/util/ImageSetConverter.java:
```diff
@@ -0,0 +1,188 @@
+            Map<String,Object> map= new HashMap<>();
+            map.putAll(dataObjectToMap(obj,dataTypes ));
+            retList.add( ImageMasterDataEntry.makeFromMap(map));
+        }
```
This method is not made to be used like this.  Setters should be called or should have a list of constants and a set with each constants.  Otherwise we are going to put junk in the object.

ejoliet replied:
Yes, we already covered that yesterday.

robyww commented on src/firefly/java/edu/caltech/ipac/firefly/util/ImageSetConverter.java:
```diff
@@ -0,0 +1,188 @@
+                continue;
+            }
+
+            switch (cls.getSimpleName().toLowerCase()) {
```
We don't care about datatypes.  All data should be `<String,String>`

lznakano commented:
I added a java class which contains all the parameters.  I moved the IrsaImageSetConverter as what we discussed.  I fixed some minor issues in other files.  Please check and let me know if any additional changes are needed.

Thanks!

robyww commented:
@ejoliet @lznakano @loitly I think we need to have a meeting about this. You are working too hard and breaking the file that should be the source of truth `ImageMasterDataEntry`.

lznakano commented:
I could just use the setXXXX to do the job.  But I don't see an issue for what I was doing.  When there are more parameters added, those parameters can be set as well.

lznakano commented:
I removed the IrsaParameter class, instead, defined enum type in ImageMasterDataEntry as you suggested.  Since the iras master table does not have imageId,  I made it up in the source code. Please take a look and let me know if any further changes are needed.  Thanks!
